### ISSUE TITLE
Optional closure 

### ISSUE THREAD

**Issue by jsaguiar** [SwiftKit/Cuckoo#12]
How can I mock a function that have an optional closure?

```swift
func optinalClosure(optionalClosure: (String -> ())?)
```

I get this error:

`Value of type 'Stubbing' (aka 'MockTestedProtocol.__StubbingProxy_TestedProtocol') has no member 'MatchedType'`

**Comment by TadeasKriz** [comment]
We need to add better support for matching optionals, which is something we've missed. Anyway, you can do a quick workaround to get it working.

If you want to mock calling the method with `nil` closure, just add this function to your code:
```
func equalToNil<T>() -> AnyMatcher<T?> {
    return equalTo(nil) { (lhs, rhs) -> Bool in
        switch (lhs, rhs) {
        case (.None, .None):
            return true
        default:
            return false
        }
    }
}
```

and then you can do `mock.optinalClosure(equalToNil())`. When you want to use `anyClosure()`, it is a bit more difficult. I've written this function for you, excuse the ugliness:

```
func anyNillableClosure<IN, OUT>() -> AnyMatcher<(IN -> OUT)?> {
    let compare: (IN -> OUT)? -> Bool = {
        if case .None = $0 { return false } else { return true }
    }
    return FunctionMatcher<(IN -> OUT)?>(function: compare, describeTo: { _ in }).typeErased()
}
```

And you use it the same way as the `equalToNil` one. Unfortunately you cannot mock for specific closures as Swift does not allow for comparison of closures.

**Comment by jsaguiar** [comment]
Thank for the help.
I don't know if it's related, but I am not able to use:

```swift
verify(mock, times(0)). optinalClosure(anyNillableClosure())
```
it's seams to be in a infinite loop.

**Comment by TadeasKriz** [comment]
This works for me. Could you show me more of the test code that leads to the infinite loop? Thanks.

**Comment by TadeasKriz** [comment]
@FilipDolnik told me there really is an infinite loop and I'm working on fixing it ASAP.

**Comment by TadeasKriz** [comment]
Fixed in Cuckoo 0.5.0.
